### ISSUE TITLE
veo_read_mem() and veo_write_mem() are not thread-safe

### ISSUE THREAD

**Issue by veos-sxarr** [SX-Aurora/aveo#16]
veo_read_mem() and veo_write_mem() are not thread-safe.
When multi threads invoke them,  "ERROR: operator()() mismatch" occurs.

Test program:

~~~
#include <stdio.h>
#include <stdlib.h>
#include <ve_offload.h>
#include <pthread.h>
#include <unistd.h>
#include <sys/syscall.h>

#define SIZE (10*1024*1024)

long syscall(long n, ...);

int gettid(void)
{
  return syscall(SYS_gettid);
}

void *child(void *proc)
{
  printf("test PID = %d, tid = %d\n", getpid(), gettid());
  uint64_t bufptr;
  uint64_t *buffer = (uint64_t *)malloc(SIZE * sizeof(uint64_t));
  for(int i = 0;i < SIZE;i++)
    buffer[i] = i;

  while (1) {
    int ret = veo_alloc_mem(proc, &bufptr, SIZE * sizeof(uint64_t));
    if (ret != 0) {
      printf("veo_alloc_mem() failed!\n");
      exit(1);
    }
    /*
    ret = veo_write_mem(proc, bufptr, buffer, SIZE * sizeof(uint64_t));
    if(ret != 0) {
      printf("veo_write_mem() failed!\n");
      exit(1);
    }
    */
    ret = veo_read_mem(proc, buffer, bufptr, SIZE * sizeof(uint64_t));
    if(ret != 0) {
      printf("veo_read_mem() failed!\n");
      exit(1);
    }
    int free_mem = veo_free_mem(proc, bufptr);
    if ( free_mem != 0) {
      printf("veo_free_mem() failed!\n");
      exit(1);
    }
  }
  free(buffer);
}


int
main()
{
  printf("main PID = %d, tid = %d\n", getpid (), gettid ());
  pthread_t th1, th2, th3, th4;
  struct veo_proc_handle *proc = veo_proc_create(-1);
  pthread_create(&th1, NULL, child, proc);
  pthread_create(&th2, NULL, child, proc);
  pthread_create(&th3, NULL, child, proc);
  pthread_create(&th4, NULL, child, proc);
  pthread_join(th1, NULL);//(void **)&proc);
  pthread_join(th2, NULL);//(void **)&proc);
  pthread_join(th3, NULL);//(void **)&proc);
  pthread_join(th4, NULL);//(void **)&proc);
  printf("end\n");
  veo_proc_destroy(proc);

  return 0;
}
~~~

Result:
~~~
$ gcc test_child.c -I/opt/nec/ve/veos/include -L/opt/nec/ve/veos/lib64 -Wl,-rpath=/opt/nec/ve/veos/lib64 -lveo -pthread
$ VE_OMP_NUM_THREADS=1 ./a.out
main PID = 31393, tid = 31393
test PID = 31393, tid = 31437
test PID = 31393, tid = 31438
test PID = 31393, tid = 31439
test PID = 31393, tid = 31440
[VH] ERROR: operator()() mismatch: dst=153e06fff010 sent_dst=601017000040
debug with : gdb -p 31393
[VH] ERROR: _progress_nolock() Internal error on executing a command(-1)
veo_read_mem() failed!
veo_read_mem() failed!
~~~

**Comment by efocht** [comment]
Thanks a lot for the report and the reproducer. The issue is now understood and I am working on a fix. It comes from the synchronous operations `veo_alloc_mem()` and `veo_free_mem()` which get mixed up with the memory transfers.

**Comment by efocht** [comment]
Solved by merge/commit 5e20b32.
Added the reproducer (slightly modified) as new test case.
Closing.
